### PR TITLE
add ephemeral procedure

### PR DESCRIPTION
**Pull request by shahidghafoor00**

### CODE REVIEW

File: cmd/wick/main.go [hunk 1]
```diff
@@ -69,6 +69,7 @@ var (
 	registerProcedure = register.Arg("procedure", "procedure name").Required().String()
 	onInvocationCmd   = register.Arg("command", "Shell command to run and return it's output").String()
 	delay             = register.Flag("delay", "Register procedure after delay (in seconds)").Int()
+	invokeCount       = register.Flag("invoke-count", "Invoke session after its called N times").Int()
```
Comment (om26er) [inline cmd/wick/main.go:72]: The text should read something like "leave session after it's called requested times"

File: wamp/main.go [hunk 2]
```diff
@@ -270,9 +270,25 @@ func Publish(session *client.Client, topic string, args []string, kwargs map[str
 	}
 }
 
-func Register(session *client.Client, procedure string, command string, delay int) {
+func Register(session *client.Client, procedure string, command string, delay int, invokeCount int) {
+
+	var isInvokeEnabled = false
```
Comment (om26er) [inline wamp/main.go:275]: we need a better name for that variable. `hasMaxInvokeCount`

File: wamp/main.go [hunk 3]
```diff
@@ -270,9 +270,25 @@ func Publish(session *client.Client, topic string, args []string, kwargs map[str
...
+
+	if invokeCount > 0 {
+		*&isInvokeEnabled = true
```
Comment (om26er) [inline wamp/main.go:278]: No need for `*&`

File: wamp/main.go [hunk 4]
```diff
@@ -270,9 +270,25 @@ func Publish(session *client.Client, topic string, args []string, kwargs map[str
...
+	}
+
 	eventHandler := func(ctx context.Context, inv *wamp.Invocation) client.InvokeResult {
 
+		if isInvokeEnabled {
```
Comment (om26er) [inline wamp/main.go:283]: This code block should executed at the end of this function not at the start

File: wamp/main.go [hunk 5]
```diff
@@ -270,9 +270,25 @@ func Publish(session *client.Client, topic string, args []string, kwargs map[str
...
+			invokeCount--
+			if invokeCount == 0 {
+				session.Done()
+				logger.Println("session closing")
+				os.Exit(1)
```
Comment (om26er) [inline wamp/main.go:288]: exit code 1 means that the program exited with an error. The error code here should be 0, which means the program exited without an error.

Comment (om26er) [reply wamp/main.go:288]: Also I am not sure if you need to exit the session manually if you already called `session.Done()`

**Review by om26er — CHANGES_REQUESTED** [review]

File: wamp/main.go (hunk 5 above)
Comment (shahidghafoor00) [reply wamp/main.go:288]: `session.Done()` didn’t exited out so I manually exited out.

File: wamp/main.go [hunk 6]
```diff
@@ -270,7 +270,14 @@ func Publish(session *client.Client, topic string, args []string, kwargs map[str
 	}
 }
 
-func Register(session *client.Client, procedure string, command string, delay int) {
+func Register(session *client.Client, procedure string, command string, delay int, invokeCount int) {
+
+	hasMaxInvokeCount := false
```
Comment (om26er) [inline wamp/main.go:275]: This code block can be simplified as
```
hasMaxInvokeCount := invokeCount > 0
```

File: wamp/main.go [hunk 7]
```diff
@@ -284,7 +291,23 @@ func Register(session *client.Client, procedure string, command string, delay in
 			return client.InvokeResult{Args: wamp.List{out}}
 		}
 
+		if hasMaxInvokeCount {
+			invokeCount--
+			if invokeCount == 0 {
```
Comment (om26er) [inline wamp/main.go:296]: once the invoke count is reached, we should just call session.Unregister(), that will ensure no other client is able to call that procedure again. Doing that essentially reduces the hacks in our logic

File: wamp/main.go [hunk 8]
```diff
@@ -284,7 +291,23 @@ func Register(session *client.Client, procedure string, command string, delay in
...
+				// FIXME This is really a hack because
+				//  there is no guarantee that the caller
+				//  has actually received the result
+				time.AfterFunc(1*time.Second, func() {
+					logger.Println("after time called")
```
Comment (om26er) [inline wamp/main.go:301]: these logs can be removed, not needed

**Review by om26er — CHANGES_REQUESTED** [review]

**Comment by om26er** [comment]
LGTM

**Pull request merged**